![](https://cbic.org.br/wp-content/uploads/2022/02/909fa5d1543363d809512626531cff03.jpeg)

# Valores Ausentes

Prof.: Nasser Santiago Boan
___

Limpeza de dados é o processo de encontrar e corrigir os dados imprecisos/incorretos que podem ocorrer no dataset. Parte desse processo é identificar, entender e tratar valores ausentes. Na vida real, a maioria dos datasets terão muitos valores ausentes e eles tem a capacidade de colocar em xeque uma análise que, em outra situação, seria considerada completa.

Por que você precisa preencher os dados ausentes? Porque a maioria dos modelos de aprendizado de máquina que você deseja usar fornecerá um erro se você passar valores ```NaN``` para ele. A maneira mais fácil é apenas preenchê-los com 0, mas isso pode reduzir significativamente a precisão do seu modelo. Para preencher os valores ausentes existem muitos métodos disponíveis. Para escolher o melhor método, você precisa entender o tipo de valor ausente e seu significado, antes de começar a preencher/excluir os dados.

## Importação
___

Utilizaremos a biblioteca ```pandas``` para fazer as análises desse dataset.

In [3]:
import pandas as pd

## Dados
___

![](https://dnn65p9ixwrwn.cloudfront.net/uploads/2022/06/logo-olist-site.png)

Os dados desse material veem da Olist, um marketplace brasileiro e que divulgou seus dados no kaggle para que a comunidade pudesse trabalhar, analisar e montar modelos. Os dados podem ser baixados através do [link](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce?select=olist_sellers_dataset.csv). Porém o arquivo zip já se encontra na pasta ```dados```, basta fazer a extração para começar o trabalho.

In [4]:
df = (
    pd.read_csv('dados/olist/olist_orders_dataset.csv')
    .merge(
        pd.read_csv('dados/olist/olist_order_reviews_dataset.csv'),
        on='order_id',
        how='left')
)

df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2018-08-08 18:37:50  
2     2018-08-22 19:07:58  
3     2017-12-05 19:21:58  
4     2018-02-18 13:02:51

## Identificando valores ausentes
___

O primeiro passo de qualquer processo de limpeza de dados é procurar valores ausentes. No ```pandas``` é possível encontrar valores ausentes de várias formas.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99992 entries, 0 to 99991
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   order_id                       99992 non-null  object 
 1   customer_id                    99992 non-null  object 
 2   order_status                   99992 non-null  object 
 3   order_purchase_timestamp       99992 non-null  object 
 4   order_approved_at              99831 non-null  object 
 5   order_delivered_carrier_date   98199 non-null  object 
 6   order_delivered_customer_date  97005 non-null  object 
 7   order_estimated_delivery_date  99992 non-null  object 
 8   review_id                      99224 non-null  object 
 9   review_score                   99224 non-null  float64
 10  review_comment_title           11568 non-null  object 
 11  review_comment_message         40977 non-null  object 
 12  review_creation_date           99224 non-null 

O primeiro método é o ```.info()``` que retorna a quantidade de linhas e de valores não-nulos. Dessa forma, basta entender que em um dataset de 99992 entradas se o método retornar que uma coluna possui um valor abaixo desse número de valores não-nulos, significa que aquela coluna tem valores nulos. Veja a coluna ```review_comment_title``` que possui 11568 valores não-nulos, ou seja, 88424 valores nulos (99992 - 11568).

In [6]:
df.isna()

order_id  customer_id  order_status  order_purchase_timestamp  \
0         False        False         False                     False   
1         False        False         False                     False   
2         False        False         False                     False   
3         False        False         False                     False   
4         False        False         False                     False   
...         ...          ...           ...                       ...   
99987     False        False         False                     False   
99988     False        False         False                     False   
99989     False        False         False                     False   
99990     False        False         False                     False   
99991     False        False         False                     False   

       order_approved_at  order_delivered_carrier_date  \
0                  False                         False   
1                  False                         False   
2                  False                         False   
3                  False                         False   
4                  False                         False   
...                  ...                           ...   
99987              False                         False   
99988              False                         False   
99989              False                         False   
99990              False                         False   
99991              False                         False   

       order_delivered_customer_date  order_estimated_delivery_date  \
0                              False                          False   
1                              False                          False   
2                              False                          False   
3                              False                          False   
4                              False                          False   
...                              ...                            ...   
99987                          False                          False   
99988                          False                          False   
99989                          False                          False   
99990                          False                          False   
99991                          False                          False   

       review_id  review_score  review_comment_title  review_comment_message  \
0          False         False                  True                   False   
1          False         False                 False                   False   
2          False         False                  True                    True   
3          False         False                  True                   False   
4          False         False                  True                    True   
...          ...           ...                   ...                     ...   
99987      False         False                  True                    True   
99988      False         False                  True                   False   
99989      False         False                  True                   False   
99990      False         False                  True                   False   
99991      False         False                  True                    True   

       review_creation_date  review_answer_timestamp  
0                     False                    False  
1                     False                    False  
2                     False                    False  
3                     False                    False  
4                     False                    False  
...                     ...                      ...  
99987                 False                    False  
99988                 False                    False  
99989                 False                    False  
99990                 False                    False  
99991                 Fals

O método ```.isna()``` performa uma operação lógica em cada célula de cada coluna do dataset e verifica se aquele valor é nulo.Contudo, somente essa operação não nos dá o resultado esperado (contar quantos valores nulos existem no dataset).

In [7]:
df.isna().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  161
order_delivered_carrier_date      1793
order_delivered_customer_date     2987
order_estimated_delivery_date        0
review_id                          768
review_score                       768
review_comment_title             88424
review_comment_message           59015
review_creation_date               768
review_answer_timestamp            768
dtype: int64

Colocando o método ```.sum()``` ao final do anterior estamos somando todos os valores de todas as colunas e o pandas retorna o resultado. Tratando o ```False``` como 0 e ```True``` como 1 o somatório nos retorna a quantidade de valores ```True```, ou seja, a quantidade de valores nulos. Também é possível utilizar o ```.isna()``` para filtragem.

In [8]:
df[df.review_comment_title.isna()]

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5      a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
...                                 ...                               ...   
99987  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99988  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99989  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99990  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99991  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5        delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
...            ...                      ...                  ...   
99987    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99988    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99989    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99990    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99991    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
5              2017-07-11 14:58:04           2017-07-26 10:57:55   
...                            ...                           ...   
99987          2017-03-10 11:18:03           2017-03-17 15:08:01   
99988          2018-02-07 23:22:42           2018-02-28 17:37:56   
99989          2017-08-28 20:52:26           2017-09-21 11:24:17   
99990          2018-01-12 15:35:03           2018-01-25 23:32:54   
99991          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                         review_id  \
0               2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
2               2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3               2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4               2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   
5               2017-08-01 00:00:00  89b738e70a1ce346db29a20fb2910161   
...                             ...                               ...   
99987           2017-03-28 00:00:00  e262b3f92d1ce917aa412a9406cf61a6   
99988           2018-03-02 00:00:00  29bb71b2760d0f876dfa178a76bc4734   
99989           2017-09-27 00:00:00  371579771219f6db2d830d50805977bb   
99990           2018-02-15 00:00:00  8ab6855b9fe9b812cd03a480a25058a1   
99991           2018-04-03 00:00:00  dc9c59b4688062c25758c2be4cafc523   

       review_score review_comment_title  \
0               4.0                  NaN   
2               5.0                  NaN   
3               5.0                  NaN   
4               5.0                  NaN   
5               4.0                  NaN   
...             ...                  ...   
99987           5.0                  NaN   
99988           4.0                  NaN   
99989           5.0                  NaN   
99990           2.0                  NaN   
99991           5.0                  NaN   

                                  review_comment_message review_cr

In [11]:
(df.isna().sum() / df.shape[0])*100

order_id                          0.000000
customer_id                       0.000000
order_status                      0.000000
order_purchase_timestamp          0.000000
order_approved_at                 0.161013
order_delivered_carrier_date      1.793143
order_delivered_customer_date     2.987239
order_estimated_delivery_date     0.000000
review_id                         0.768061
review_score                      0.768061
review_comment_title             88.431074
review_comment_message           59.019722
review_creation_date              0.768061
review_answer_timestamp           0.768061
dtype: float64

Agora que conseguimos identificar os valores ausentes, o que fazer? Existem duas possibilidades:

1. Dropar os valores nulos

    *Dropar* os valores nulos significa removê-los do dataset por completo. Esse processo pode ser feito em colunas inteiras ou em linhas específicas, o pandas possui métodos para realizar esse trabalho.

2. Substituir os valores nulos

    O processo de substitução dos valores nulos muitas vezes é chamado de *imputação* (*imputation* em inglês) e consiste em adicionar valores ao dataset escolhidos através de heurística.

## Dropando valores nulos

In [17]:
## dropando as colunas que possuem pelo menos um valor nulo

df.dropna(axis=1)

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99987  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99988  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99989  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99990  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99991  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp order_estimated_delivery_date  
0        delivered      2017-10-02 10:56:33           2017-10-18 00:00:00  
1        delivered      2018-07-24 20:41:37           2018-08-13 00:00:00  
2        delivered      2018-08-08 08:38:49           2018-09-04 00:00:00  
3        delivered      2017-11-18 19:28:06           2017-12-15 00:00:00  
4        delivered      2018-02-13 21:18:39           2018-02-26 00:00:00  
...            ...                      ...                           ...  
99987    delivered      2017-03-09 09:54:05           2017-03-28 00:00:00  
99988    delivered      2018-02-06 12:58:58           2018-03-02 00:00:00  
99989    delivered      2017-08-27 14:46:43           2017-09-27 00:00:00  
99990    delivered      2018-01-08 21:28:27           2018-02-15 00:00:00  
99991    delivered      2018-03-08 20:57:30           2018-04-03 00:00:00  

[99992 rows x 5 columns]

In [18]:
df.dropna(axis=1).isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_estimated_delivery_date    0
dtype: int64

In [26]:
## dropando as colunas com 80% ou mais de valores nulos (explicar passo a passo)

df.drop(df.columns[(df.isna().sum() / df.shape[0]) >= 0.8],axis=1)

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99987  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99988  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99989  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99990  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99991  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99987    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99988    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99989    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99990    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99991    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99987          2017-03-10 11:18:03           2017-03-17 15:08:01   
99988          2018-02-07 23:22:42           2018-02-28 17:37:56   
99989          2017-08-28 20:52:26           2017-09-21 11:24:17   
99990          2018-01-12 15:35:03           2018-01-25 23:32:54   
99991          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                         review_id  \
0               2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1               2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2               2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3               2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4               2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   
...                             ...                               ...   
99987           2017-03-28 00:00:00  e262b3f92d1ce917aa412a9406cf61a6   
99988           2018-03-02 00:00:00  29bb71b2760d0f876dfa178a76bc4734   
99989           2017-09-27 00:00:00  371579771219f6db2d830d50805977bb   
99990           2018-02-15 00:00:00  8ab6855b9fe9b812cd03a480a25058a1   
99991           2018-04-03 00:00:00  dc9c59b4688062c25758c2be4cafc523   

       review_score                             review_comment_message  \
0               4.0  Não testei o produto ainda, mas ele veio corre...   
1               4.0                               Muito bom o produto.   
2               5.0                                                NaN   
3               5.0  O produto foi exatamente o que eu esperava e e...   
4               5.0                                                NaN   
...             ...                                                ...   
99987           5.0                                                NaN   
999

Perceba que até então nenhuma das operações que fizemos alterou o dataset original. Para "salvar" as alterações feitas até então usamos o artfício de refazer a operação de atribuição (usando símbolo '='), algumas funções pandas possuem um argumento chamado ```inplace``` que funciona da seguinte forma:

![](imagens/inplace.png)

In [27]:
df.shape

(99992, 14)

In [31]:
df.dropna(axis=0,how='any')

order_id                       customer_id  \
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
15     dcb36b511fcac050b97cd5c05de84dc3  3b6828a50ffe546942b7a473d70ac0fc   
24     f3e7c359154d965827355f39d6b1fdac  62b423aab58096ca514ba6aa06be2f98   
26     acce194856392f074dbf9dada14d8d82  7e20bf5ca92da68200643bda76c504c6   
31     1790eea0b567cf50911c057cf20f90f9  52142aa69d8d0e1247ab0cada0f76023   
...                                 ...                               ...   
99912  c7510fe2bce3ab54854dc588ce4d187c  f74562da630b57572d1742ba2f28b662   
99920  52018484704db3661b98ce838612b507  e450a297a7bc6839ceb0cf1a2377fa02   
99928  a6bd1f93b7ff72cc348ca07f38ec4bee  6d63fa86bd2f62908ad328325799152f   
99946  5597332b7eded552f104108f22b023e4  aaa423fb52f4106f477683490cbd5845   
99978  19dba6d1e58b04e7fc820558a38ea398  cf97e9cf454480b303de6736ddd15fd6   

      order_status order_purchase_timestamp    order_approved_at  \
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
15       delivered      2018-06-07 19:03:12  2018-06-12 23:31:02   
24       delivered      2018-08-09 11:44:40  2018-08-10 03:24:51   
26       delivered      2018-06-04 00:00:13  2018-06-05 00:35:10   
31       delivered      2018-04-16 21:15:39  2018-04-16 22:10:26   
...            ...                      ...                  ...   
99912    delivered      2018-05-16 09:05:27  2018-05-17 09:18:23   
99920    delivered      2018-08-29 12:25:59  2018-08-29 12:35:17   
99928    delivered      2018-04-20 17:28:40  2018-04-24 19:26:10   
99946    delivered      2018-08-15 13:03:37  2018-08-15 13:15:22   
99978    delivered      2018-05-12 13:42:24  2018-05-12 13:55:17   

      order_delivered_carrier_date order_delivered_customer_date  \
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
15             2018-06-11 14:54:00           2018-06-21 15:34:32   
24             2018-08-10 12:29:00           2018-08-13 18:24:27   
26             2018-06-05 13:24:00           2018-06-16 15:20:55   
31             2018-04-18 13:05:09           2018-05-05 12:28:34   
...                            ...                           ...   
99912          2018-05-18 14:23:00           2018-05-22 14:20:48   
99920          2018-08-29 13:38:00           2018-08-30 22:48:27   
99928          2018-04-23 17:18:40           2018-04-28 17:38:42   
99946          2018-08-15 14:37:00           2018-08-17 16:06:37   
99978          2018-05-16 08:14:00           2018-05-18 17:18:29   

      order_estimated_delivery_date                         review_id  \
1               2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
15              2018-07-04 00:00:00  5a8608437f4b52bec5cd143f37258b5e   
24              2018-08-17 00:00:00  9a55cbc65c8fad47581da5ebbb2b373e   
26              2018-07-18 00:00:00  2d4303fd561eb4ab51c87faa6598dc10   
31              2018-05-15 00:00:00  764c587207b9589949abd863bdb72dec   
...                             ...                               ...   
99912           2018-06-21 00:00:00  bbe7c6000965c0d7b0a0fbd9c9fa7b24   
99920           2018-09-03 00:00:00  7a11bf826668febba0800ec35884958c   
99928           2018-05-15 00:00:00  eb4c94cf604d9cf4d4442890ac33797d   
99946           2018-08-21 00:00:00  c44004d7e60dc281ebd5361b717570c3   
99978           2018-06-04 00:00:00  d842b3001e953f4ca2a3a312d02523f9   

       review_score review_comment_title  \
1               4.0     Muito boa a loja   
15              5.0             Nota dez   
24              5.0               Ótimo    
26              1.0   NÃO RECOMENDO!!!!    
31              1.0          Enganadores   
...             ...                  ...   
99912           5.0           excelente    
99920           1.0     Muito frágil !!!   
99928           5.0            recomendo   
99946           5.0                Ótima   
99978           5.0        Bom.maravilha   

                                  review_comment_message review_cr

In [35]:
df.dropna(how='any',subset=['review_score','customer_id','order_delivered_customer_date','order_approved_at',]).isna().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                    0
order_delivered_carrier_date         1
order_delivered_customer_date        0
order_estimated_delivery_date        0
review_id                            0
review_score                         0
review_comment_title             85141
review_comment_message           57252
review_creation_date                 0
review_answer_timestamp              0
dtype: int64

Também é possível fazer o trabalho de imputação de valores nulos. Existem várias técnicas para fazer imputação e até modelos de machine learning para executar esse tipo de trabalho, porém vamos tratar somente dos métodos mais simples.

In [36]:
df.isna().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  161
order_delivered_carrier_date      1793
order_delivered_customer_date     2987
order_estimated_delivery_date        0
review_id                          768
review_score                       768
review_comment_title             88424
review_comment_message           59015
review_creation_date               768
review_answer_timestamp            768
dtype: int64

## Imputando valores nulos

O métodos ```.fillna()``` encontra os valores nulos de uma série ou de todo o dataframe e substitui pelo valor passado como argumento.

In [40]:
df.review_score.fillna(df.review_score.mean())

0        4.0
1        4.0
2        5.0
3        5.0
4        5.0
        ... 
99987    5.0
99988    4.0
99989    5.0
99990    2.0
99991    5.0
Name: review_score, Length: 99992, dtype: float64

In [41]:
df.review_score.fillna(df.review_score.mean()).isna().sum()

0

In [48]:
df.review_score.describe()

count    99224.000000
mean         4.086421
std          1.347579
min          1.000000
25%          4.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: review_score, dtype: float64

In [50]:
df.review_score.fillna(df.review_score.mean()).describe()

count    99992.000000
mean         4.086421
std          1.342394
min          1.000000
25%          4.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: review_score, dtype: float64

### Exercício
#### 1. Qual o tempo médio entre aprovação da compra até a entrega para cada nota de pedido?
#### 2. Quais são as palavras mais utilizadas na coluna "review_comment_message" para cada nota de pedido?
#### 3. Use os datasets de pedidos, reviews, clientes e pagamentos para responder:
    a. Qual estado do país fez mais pedidos?
    b. Qual a média de nota por estado?
    c. Qual o ticket médio por estado?
    d. Qual o método de pagamento preferido dos estados do norte do Brasil?
    e. Qual o método de pagamento preferido dos estados do sudeste do Brasil?